### ISSUE TITLE
SUPPORT VECTOR MACHINE

### ISSUE THREAD

**Issue by AHAN07-svg** [AHAN07-svg/Support-vector-machine#1]
import pandas as pd

import numpy as np

from sklearn.preprocessing import LabelEncoder

from sklearn.preprocessing import MinMaxScaler

from sklearn.model_selection import train_test_split

from sklearn.metrics import classification_report, accuracy_score

from sklearn.svm import SVC

import matplotlib.pyplot as plt

from sklearn.model_selection import learning_curve

import seaborn as sns

from sklearn.decomposition import PCA

# Load the dataset

housing_df = pd.read_csv("californiahousing.csv")

# Display basic information

print(housing_df.head())

print(housing_df.info())

print(housing_df.shape)

# Handle missing values

housing_df = housing_df.dropna()

# Encode categorical variables if any

label_encoder = LabelEncoder()

for col in housing_df.select_dtypes(include=["object"]).columns:

    housing_df[col] = label_encoder.fit_transform(housing_df[col])

# Define features and target

X = housing_df.drop("median_house_value", axis=1)

y = housing_df["median_house_value"]

# Convert target into categories (classification)

y = pd.qcut(y, q=2, labels=[0, 1])

# Feature scaling

scaler = MinMaxScaler()

X_scaled = scaler.fit_transform(X)

# Train-test split

X_train, X_test, y_train, y_test = train_test_split(

    X_scaled, y, test_size=0.2, random_state=42

)

# Train SVM model with RBF kernel

svc_rbf = SVC(kernel="rbf", gamma="scale")

svc_rbf.fit(X_train, y_train)

# Predictions

y_pred = svc_rbf.predict(X_test)

# Evaluation

print("Accuracy:", accuracy_score(y_test, y_pred))

print(classification_report(y_test, y_pred))

# Learning curve

train_sizes, train_scores, test_scores = learning_curve(

    svc_rbf, X_train, y_train, cv=5

)

train_scores_mean = np.mean(train_scores, axis=1)

test_scores_mean = np.mean(test_scores, axis=1)

plt.plot(train_sizes, train_scores_mean, label="Training score")

plt.plot(train_sizes, test_scores_mean, label="Cross-validation score")

plt.xlabel("Training Size")

plt.ylabel("Score")

plt.title("Learning Curve for SVM")

plt.legend()

plt.show()

# PCA for visualization

pca = PCA(n_components=2)

X_train_pca = pca.fit_transform(X_train)

X_test_pca = pca.transform(X_test)

support_vectors_pca = pca.transform(svc_rbf.support_vectors_)

plt.scatter(X_train_pca[:, 0], X_train_pca[:, 1], c=y_train, cmap="coolwarm", alpha=0.6)

plt.scatter(

    support_vectors_pca[:, 0],

    support_vectors_pca[:, 1],

    color="black",

    marker="x",

    label="Support Vectors"

)

plt.xlabel("Principal Component 1")

plt.ylabel("Principal Component 2")

plt.title("Support Vectors in SVM (RBF Kernel)")

plt.legend()

plt.show()
